aten/src/ATen/cuda/CUDASparseDescriptors.h

```diff
@@ -0,0 +1,202 @@
+#pragma once
+
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDADataType.h>
+#include <ATen/cuda/CUDASparse.h>
+#include <ATen/native/LinearAlgebraUtils.h>
+#include <ATen/native/cuda/MiscUtils.h>
+
+#include <c10/core/ScalarType.h>
+
+#if AT_USE_CUSPARSE_GENERIC_API()
+
+namespace at {
+namespace cuda {
+namespace sparse {
+
+namespace {
+
+// If a specific GPU model does not provide native support for a given data
+// type, cuSparse routines return CUSPARSE_STATUS_ARCH_MISMATCH error
+void check_supported_cuda_type(cudaDataType cuda_type) {
+  if (cuda_type == CUDA_R_16F) {
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    TORCH_CHECK(
+        prop->major >= 5 && ((10 * prop->major + prop->minor) >= 53),
+        "Sparse operations with CUDA tensors of Float16 type are not supported on GPUs with compute capability < 5.3 (current: ",
+        prop->major,
+        ".",
+        prop->minor,
+        ")");
+  }
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11000
+  if (cuda_type == CUDA_R_16BF) {
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    TORCH_CHECK(
+        prop->major >= 8,
+        "Sparse operations with CUDA tensors of BFloat16 type are not supported on GPUs with compute capability < 8.0 (current: ",
+        prop->major,
+        ".",
+        prop->minor,
+        ")");
+  }
+#endif
+}
+
+} // anonymous namespace
+
+inline cusparseIndexType_t getCuSparseIndexType(const c10::ScalarType& scalar_type) {
+  if (scalar_type == c10::ScalarType::Int) {
+    return CUSPARSE_INDEX_32I;
+  } else if (scalar_type == c10::ScalarType::Long) {
+    return CUSPARSE_INDEX_64I;
+  } else {
+    TORCH_INTERNAL_ASSERT(
+        false,
+        "Cannot convert type ",
+        scalar_type,
+        " to cusparseIndexType.");
+  }
+}
+
+template <typename T, cusparseStatus_t (*destructor)(T*)>
+struct CuSparseDescriptorDeleter {
+  void operator()(T* x) {
+    if (x != nullptr) {
+      TORCH_CUDASPARSE_CHECK(destructor(x));
+    }
+  }
+};
+
+template <typename T, cusparseStatus_t (*destructor)(T*)>
+class CuSparseDescriptor {
+ public:
+  T* descriptor() const {
+    return descriptor_.get();
+  }
+  T* descriptor() {
+    return descriptor_.get();
+  }
+
+ protected:
+  std::unique_ptr<T, CuSparseDescriptorDeleter<T, destructor>> descriptor_;
+};
+
+class TORCH_CUDA_CPP_API CuSparseDnMatDescriptor
+    : public CuSparseDescriptor<cusparseDnMatDescr, &cusparseDestroyDnMat> {
+ public:
+  CuSparseDnMatDescriptor(const Tensor& input) {
+    TORCH_INTERNAL_ASSERT(input.layout() == kStrided);
```

Comment:
Is there value in using TORCH_CHECK here instead or are you already guarding this with more user friendly error messages at the use sites?

Reply:
This class is never exposed to the Python user in any way, so all those asserts are only for developers to state the assumptions on the input tensor.